### ISSUE TITLE
Support Node 18

### ISSUE THREAD

**Issue by mhofman** [Agoric/Agoric#6489]
## What is the Problem Being Solved?

Node 18 is [reaching LTS imminently](https://github.com/nodejs/Release#release-schedule). We should make sure everything works when using node 18.

## Description of the Design

Setup a dev env with node 18 and see what breaks
Update CI to also run against node 18.

## Security Considerations
None

## Test Plan
Yes

**Comment by sigv** [comment]
When trying to run the `pismoA` tag with Node.js 18.x we are observing an initialization error:

```none
launch-chain: Launching SwingSet kernel
portHandler threw (Error#1)
Error#1: You must provide the URL of lib/mappings.wasm by calling SourceMapConsumer.initialize({ 'lib/mappings.wasm': ... }) before using SourceMapConsumer
  at async Promise.all (index 0)
  at async Promise.all (index 6)
  at async allValues (packages/SwingSet/src/controller/initializeSwingset.js:29:30)
  at async buildVatAndDeviceBundles (packages/SwingSet/src/controller/initializeSwingset.js:54:19)
  at async initializeSwingset (packages/SwingSet/src/controller/initializeSwingset.js:336:25)
  at async ensureSwingsetInitialized (packages/cosmic-swingset/src/launch-chain.js:113:5)
  at async buildSwingset (packages/cosmic-swingset/src/launch-chain.js:115:3)
  at async launch (packages/cosmic-swingset/src/launch-chain.js:264:52)
  at async launchAndInitializeSwingSet (packages/cosmic-swingset/src/chain-main.js:412:15)
  at async toSwingSet (packages/cosmic-swingset/src/chain-main.js:461:22)
  at async portHandlers.<computed> (packages/cosmic-swingset/src/chain-main.js:159:16)
Cannot initialize Controller Error: You must provide the URL of lib/mappings.wasm by calling SourceMapConsumer.initialize({ 'lib/mappings.wasm': ... }) before using SourceMapConsumer
Error: exit status 1
```

This seems to be caused by [mozilla/source-map#454](https://github.com/mozilla/source-map/issues/454#issuecomment-1147637420), fixed with [`source-map@0.7.4`](https://www.npmjs.com/package/source-map/v/0.7.4).
[`source-map@0.7.3`](https://www.npmjs.com/package/source-map/v/0.7.3) is in place in repository:

```none
$ npm ls source-map
@agoric/sdk@ /home/agoric/source/agoric-upgrade-8
└─┬ @agoric/cosmic-swingset@0.39.2 -> ./packages/cosmic-swingset
  ├─┬ @agoric/deploy-script-support@0.9.4 -> ./packages/deploy-script-support
  │ └─┬ @endo/bundle-source@2.3.1
  │   ├─┬ @agoric/babel-generator@7.17.6
  │   │ └── source-map@0.5.7
  │   └── source-map@0.7.3
  ├─┬ @agoric/xsnap@0.13.2 -> ./packages/xsnap
  │ └─┬ rollup-plugin-terser@5.3.1
  │   └─┬ terser@4.8.0
  │     ├─┬ source-map-support@0.5.21
  │     │ └── source-map@0.6.1 deduped
  │     └── source-map@0.6.1
  ├─┬ agoric@0.18.2 -> ./packages/agoric-cli
  │ └─┬ dd-trace@3.3.0
  │   └─┬ @datadog/pprof@1.0.2
  │     └── source-map@0.7.3
  └─┬ c8@7.11.0
    └─┬ v8-to-istanbul@8.1.1
      └── source-map@0.7.3
```

**Comment by mhofman** [comment]
Thanks for the report. We had an internal report of a similar error. For now we recommend you use Node 16 LTS with pismoA. We may consider an intermediary patch release if we can guarantee the update is compatible with the current release.

**Comment by mhofman** [comment]
@sigv according to the [upgrade instructions](https://github.com/Agoric/agoric-sdk/wiki/ag0-to-agd-upgrade-for-mainnet-1-launch#install-nodejs-and-yarn), please use Node 16. We will not be supporting Node 18 for `pismoA`.

**Comment by sigv** [comment]
> @sigv according to the [upgrade instructions](https://github.com/Agoric/agoric-sdk/wiki/ag0-to-agd-upgrade-for-mainnet-1-launch#install-nodejs-and-yarn), please use Node 16. We will not be supporting Node 18 for `pismoA`.

Understandable. Looking forward to having Current LTS supported on the next version then 🤞🏻

**Comment by dckc** [comment]
@erights reports some symptoms when trying v18.12.1:

```
yarn test test/swingsetTests/makeKind/test-makeKind.js
yarn run v1.22.19
$ ava --verbose test/swingsetTests/makeKind/test-makeKind.js

  ✘ [fail]: before hook Rejected promise returned by test
  ✘ 1 test remaining in test/swingsetTests/makeKind/test-makeKind.js
  ─

  before hook

  Rejected promise returned by test. Reason:

  Error {
    message: 'You must provide the URL of lib/mappings.wasm by calling SourceMapConsumer.initialize({ \'lib/mappings.wasm\': ... }) before using SourceMapConsumer',
  }

  › async allValues (packages/SwingSet/src/controller/initializeSwingset.js:28:30)
  › async buildVatAndDeviceBundles (packages/SwingSet/src/controller/initializeSwingset.js:53:19)
  › async buildKernelBundles (packages/SwingSet/src/controller/initializeSwingset.js:83:37)
  › async packages/zoe/test/swingsetTests/makeKind/test-makeKind.js:21:25

  ─

  1 hook failed
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by erights** [comment]
Yeah, looks like the same problem @sigv reported above at https://github.com/Agoric/agoric-sdk/issues/6489#issuecomment-1295922938

**Comment by erights** [comment]
Curiously, in the experiment at https://github.com/Agoric/agoric-sdk/pull/6549 , when I locally run the same
```sh
yarn test test/swingsetTests/makeKind/test-makeKind.js
```
under Node v18, it works.

**Comment by sigv** [comment]
With #6560 (`source-map@0.7.4`) the mentioned test case from `packages/zoe` succeeds as well.

```plain
yarn run v1.22.19
$ ava --verbose test/swingsetTests/makeKind/test-makeKind.js

bundling: 6.556s kernel, 0.171s contracts, 1.731s vats, 8.458s total
  ✔ defineKind swingset (4.3s)
  ─

  1 test passed
Done in 14.00s.
```

**Comment by erights** [comment]
@mhofman 
Now that https://github.com/Agoric/agoric-sdk/pull/6560 has been merged, can we close this issue?

**Comment by mhofman** [comment]
We need to add CI tests to ensure continuous Node 18 compatibility (and maybe a round of manual end to end testing on node 18)

**Comment by sigv** [comment]
I understand this is of fairly low priority to the team, so to keep the ball moving #6599 is available with the test suite enabling Node.js 18.x. Further manual testing can be conducted at a later time as well.

**Comment by sigv** [comment]
@mhofman, do you prefer to update [`packages/deployment/Dockerfile.sdk`](https://github.com/Agoric/agoric-sdk/blob/85d29682fc358316ee7f3993189e86d8f3c5c356/packages/deployment/Dockerfile.sdk) to also build with `node:18-bullseye` instead of current `node:16-buster`?

That _should_ be all the Node.js framework changes for pipelines/CI (until Node.js 14 is dropped off in roughly 5 months). Manual testing can be carried out to ensure compatibility, as mentioned in previous commit.

**Comment by mhofman** [comment]
I think we should keep the recommended / default to 16 right now. This IPv6 default in 18 has me worried about subtle breakage, especially with Docker.

**Comment by sigv** [comment]
Makes sense. Then it's just a question of running manual testing before deciding to officially claim Node.js 18 as compatible. Again - thank you! 🤞🏻

**Comment by dckc** [comment]
fixed in #7073 / 711059c9e8a152e545a4b12b45821e803fc6623a
